IdentityCore/src/util/NSString+MSIDExtensions.m

```diff
@@ -99,33 +99,22 @@ - (NSString *)msidTrimmedString
 - (NSString *)msidWWWFormURLDecode
 {
     // Two step decode: first replace + with a space, then percent unescape
-    CFMutableStringRef decodedString = CFStringCreateMutableCopy( NULL, 0, (__bridge CFStringRef)self );
-    CFStringFindAndReplace( decodedString, CFSTR("+"), CFSTR(" "), CFRangeMake( 0, CFStringGetLength( decodedString ) ), kCFCompareCaseInsensitive );
-    
-    CFStringRef unescapedString = CFURLCreateStringByReplacingPercentEscapes( NULL,                    // Allocator
-                                                                                          decodedString,           // Original string
-                                                                                          CFSTR("")); // Encoding
-    CFRelease( decodedString );
-    
-    return CFBridgingRelease(unescapedString);
+    return [[self stringByReplacingOccurrencesOfString:@"+" withString:@" "] stringByRemovingPercentEncoding];
 }
 
 
 - (NSString *)msidWWWFormURLEncode
 {
-    static NSCharacterSet* set = nil;
- 
-    static dispatch_once_t onceToken;
-    dispatch_once(&onceToken, ^{
-        
-        NSMutableCharacterSet *allowedSet = [[NSCharacterSet URLQueryAllowedCharacterSet] mutableCopy];
-        [allowedSet addCharactersInString:@" "];
-        [allowedSet removeCharactersInString:@"!$&'()*+,/:;=?@"];
-        
-        set = allowedSet;
-    });
-    
-    NSString *encodedString = [self stringByAddingPercentEncodingWithAllowedCharacters:set];
+    /*https://tools.ietf.org/html/rfc3986#section-3
+     Characters that are allowed in a URI but do not have a reserved
+     purpose are called unreserved.  These include uppercase and lowercase
+     letters, decimal digits, hyphen, period, underscore, and tilde.
+     unreserved  = ALPHA / DIGIT / "-" / "." / "_" / "~"
+     */
+    NSMutableCharacterSet *allowedCharacters = [NSMutableCharacterSet alphanumericCharacterSet];
+    NSString *unreserved = @"-._~ ";
```

Comment:
note: this now makes underscore a valid character while it wasn't before.
Also, creating character set only once is a nice optimization, not sure why it got removed.

Reply:
But you're doing the same - taking alphanumericCharacterSet and adding characters to it. One less operation then before, but doing it only once won't hurt. Regarding underscore, we need to print URLQueryAllowedCharacterSet and see if it's there or not.

Reply:
I have printed the URLQueryAllowedCharacterSet and confirmed that it contains '_'